scipy/optimize/tests/test_zeros.py

```diff
@@ -302,6 +315,69 @@ def test_root_scalar_fail(self):
         with pytest.raises(ValueError):
             root_scalar(f1, method='halley', fprime2=f1_2, x0=3, xtol=1e-6)  # no fprime
 
+    def test_array_chandrupatla(self):
+        """test _chandrupatla with array, similar to newton method"""
+
+        k = np.arange(1,8)
+        result = zeros._chandrupatla(self.f3, 0, 3, args=(k,))
+        x_expected = ([1., 1.41421356, 1.73205081,
+                      2., 2.23606798,2.44948974,
+                      2.64575131])
+        assert_allclose(result.x, x_expected)
+
+    def test_chandrupatla_newton_similar(self):
+        """test that _chandrupatla is giving a similar result to newton"""
+
+        k = np.arange(1,8)
+
+        result = zeros._chandrupatla(self.f3, 0, 3, args=(k[0],))
+        expected = zeros.newton(self.f3, x0=0, x1=3, args=(k[0],))
+        assert_allclose(result.x, expected)
+
+    def test_chandrupatla_brent_similar(self):
+        """test than chandrupatla gives equal scalar result to brentq"""
+        def f3(x):
+            return (x**3 - 1)
+
+        result = zeros._chandrupatla(f3,0,3,maxiter=50)
+        expected = zeros.brentq(f3,0,3)
+
+        assert_allclose(result.x, expected)
+
+    def test_chandrupatla_np_vectorize(self):
+        """test that _chandrupatla handles input with np.vectorize"""
+
+        k = np.arange(1,8)
+
+        res = zeros._chandrupatla(self.f3,0,3,
+                                  args=(k,),
+                                  maxiter=50)
+
+        vec_res = zeros._chandrupatla(np.vectorize(self.f3),0,3,
+                                  args=(k,),
+                                  maxiter=50)
+
+        assert_allclose(res.fun, vec_res.fun)
+        assert_allclose(res.x, vec_res.x)
+
```

Comment:
I'll take care of this and other tests in a PR to your branch.

Reply:
The test I'm suggesting is something like this:

```python3
import numpy as np
from numpy.testing import assert_allclose, assert_equal
from scipy import stats, optimize, integrate
import matplotlib.pyplot as plt
from mpmath import mp
mp.dps = 50
rng = np.random.default_rng(1638083107694713882823079058616272161)

from scipy.optimize._zeros_py import _chandrupatla

fields = ('message, fun, converged, flag, function_calls, iterations, root, '
          'lower_bracket, upper_bracket, lower_fun_value, upper_fun_value')
fields = fields.split(', ')

def unpack_result(res):
    unpacked = tuple([res.get(field) for field in fields])
    return unpacked

from scipy import stats
f = stats.argus(1).cdf
a, b = 0.001, 0.999
# note that f(x) cannot return -0.05 or 1.05 for any real `x`
rhs = np.linspace(-0.05, 1.05, 12)
res = _chandrupatla(lambda x: f(x) - rhs, a, b)
res = unpack_result(res)

@np.vectorize
def chandrupatla_single(rhs):
    res =  _chandrupatla(lambda x: f(x) - rhs, a, b)
    return unpack_result(res)

ref = chandrupatla_single(rhs)

for i, res_i, ref_i in zip(range(len(res)), res, ref):
    try:
        np.testing.assert_allclose(res_i, ref_i, atol=1e-10)
    except Exception as e:
        print(f'==={fields[i]}===')
        print(str(e))
```
although we won't want to use `assert_allclose` on everything. Rather, we'd check equality of messages, the function value of `res` would be less than or equal to that of `ref` (because more iterations may be performed on some elements after the tolerance is met), etc. But this is already pointing out some real failures:

```
===converged===
Not equal to tolerance rtol=1e-07, atol=1e-10
Mismatched elements: 10 / 12 (83.3%)
 x: array([False, False, False, False, False, False, False, False, False,
       False, False, False])
 y: array([False,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True, False])
===flag===
Not equal to tolerance rtol=1e-07, atol=1e-10
Mismatched elements: 10 / 12 (83.3%)
Max absolute difference: 1
Max relative difference: 0.
 x: array([-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1])
 y: array([-1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, -1])
===function_calls===
Not equal to tolerance rtol=1e-07, atol=1e-10
Mismatched elements: 12 / 12 (100%)
Max absolute difference: 20
Max relative difference: 6.66666667
 x: array([23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23])
 y: array([ 3, 11,  9,  8,  8,  8,  8,  9,  8, 12, 11,  4])
===iterations===
Not equal to tolerance rtol=1e-07, atol=1e-10
Mismatched elements: 11 / 12 (91.7%)
Max absolute difference: 15.
Max relative difference: 2.33333333
 x: array([ 2., 22., 22., 22., 21., 22., 22., 22., 22., 22., 22., 10.])
 y: array([ 2., 10.,  8.,  7.,  7.,  7.,  7.,  8.,  7., 11., 10.,  3.])

```